tempesta_fw/http.c

```diff
@@ -2341,85 +2351,92 @@ tfw_http_req_mark_error(TfwHttpReq *req, int status)
 }
 
 /**
- * Functions define logging and response behaviour during detection of
- * malformed or malicious messages. Mark client connection in special
- * manner to delay its closing until transmission of error response
- * will be finished.
+ * Function defines logging and response behaviour during detection of
+ * malformed or malicious messages. If reply to client must be sent,
+ * mark client connection in special manner to delay its closing until
+ * transmission of error response will be finished; otherwise, close
+ * connection immediately.
  */
 static void
-tfw_http_cli_error_resp_and_log(bool reply, bool nolog, TfwHttpReq *req,
-				int status, const char *msg)
+tfw_http_error_resp_and_log(bool reply, bool nolog, TfwHttpReq *req,
+			    int status, const char *msg, bool close)
 {
 	if (!nolog)
 		TFW_WARN_ADDR(msg, &req->conn->peer->addr);
 
 	if (reply) {
-		TfwCliConn *cli_conn = (TfwCliConn *)req->conn;
-		tfw_connection_unlink_msg(req->conn);
-		spin_lock(&cli_conn->seq_qlock);
-		list_add_tail(&req->msg.seq_list, &cli_conn->seq_queue);
-		spin_unlock(&cli_conn->seq_qlock);
+		if (!close) {
+			TfwCliConn *cli_conn = (TfwCliConn *)req->conn;
+			tfw_connection_unlink_msg(req->conn);
+			spin_lock(&cli_conn->seq_qlock);
+			list_add_tail(&req->msg.seq_list, &cli_conn->seq_queue);
+			spin_unlock(&cli_conn->seq_qlock);
+		}
 		tfw_http_req_mark_error(req, status);
+		return;
 	}
-	else {
-		tfw_http_conn_req_clean(req);
-	}
-}
 
-static void
-tfw_http_srv_error_resp_and_log(bool reply, bool nolog, TfwHttpReq *req,
-				int status, const char *msg)
-{
-	if (!nolog)
-		TFW_WARN_ADDR(msg, &req->conn->peer->addr);
+	if (close)
+		ss_close_sync(req->conn->sk, true);
```

Comment:
If `reply == false` and `close == true`, the client connection will be closed and all requests in it will be discarded. Probably that is not a desired behaviour for ` tfw_client_drop_close()`: we would like to normally respond to all the previous requests, and close the connection after that.  I've already done some improvements  here: https://github.com/tempesta-tech/tempesta/pull/1012/files#diff-0d15717c2e20453f4271befe0024c2d4R2484 But there still some questions to discus. I would suggest to discuss it in the chat and probably postpone the change for the future PR.

Reply:
Changes had been reverted.